### PR TITLE
hotfix to v0.3.0-rc1 restart when lock not acquired

### PR DESCRIPTION
**Pull request by lanzafame**
hotfix to v0.3.0-rc1

### CODE REVIEW

**Review by coryschwartz — APPROVED** [review]

File: schedule/scheduler.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ type TaskConfig struct {
 	// Locker is an optional lock that must be taken before the task can execute.
 	Locker Locker
 
+	// ExitOnFailure will will cause cause a fatal failure when a task stops with an error.
+	ExitOnFailure bool
```
Comment (frrist) [inline schedule/scheduler.go:34]: I'm not sure we need this flag. Setting `RestartOnFailuer` to false more or less does the same thing.

File: schedule/scheduler.go [hunk 2]
```diff
@@ -91,9 +94,15 @@ func (s *Scheduler) Run(ctx context.Context) error {
 				if err := tc.Locker.Lock(ctx); err != nil {
 					if errors.Is(err, storage.ErrLockNotAcquired) {
 						log.Infow("task not started: lock not acquired", "task", tc.Name)
+						if tc.ExitOnFailure {
+							log.Fatal("Exiting due to previous error. ExitOnFailure")
```
Comment (frrist) [inline schedule/scheduler.go:98]: I'm not sure this is the behavior we want:
```
// Fatal uses fmt.Sprint to construct and log a message, then calls os.Exit.
```
https://github.com/uber-go/zap/blob/master/sugar.go#L126

None of the deferred unlocking code will be called if we exit here, and per my above comment I don't think the `ExitOnFailure` flag is needed.

File: schedule/scheduler.go [hunk 3]
```diff
@@ -133,6 +142,9 @@ func (s *Scheduler) Run(ctx context.Context) error {
 					}
 					log.Errorw("task exited with failure", "task", tc.Name, "error", err.Error())
 
+					if tc.ExitOnFailure {
+						log.Fatal("Exiting due to previous error. ExitOnFailure")
+					}
```
Comment (frrist) [inline schedule/scheduler.go:147]: ditto

**Review by willscott — APPROVED** [review]
only looked at changes since previous approval. looks fine.

File: storage/migrations/17_chainvis_views_not_materialized.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
+package migrations
```
Comment (frrist) [inline storage/migrations/17_chainvis_views_not_materialized.go:1]: What is the motivation for calling this before the migration previously holding this number (17)? An alternative might be:
- Leave 17 (multisig transactions) and 18 (drops unused tables) alone. 
- Add migration 19 with the desired changes from here.

**Review by frrist — COMMENTED** [review]
I'd like to find an alternative to `ExitOnFailure`, I don't see why we need it.

We cannot merge this with the usage of Fatal logs, I believe they will break the locking.

File: schedule/scheduler.go (hunk 1 above)
Comment (placer14) [reply schedule/scheduler.go:34]: This fix is needed for the indexer when it fails on startup (as the RestartOnFailure flag is not considered until after locks are added which is exactly where the index fails when locks could not be acquired. (https://github.com/filecoin-project/sentinel-visor/blob/master/schedule/scheduler.go#L90-L106)

Agree it's a cludgy fix and would prefer to consider `RestartOnCompletion` inside the locking logic, but this is what was delivered and is working in production right now. Propose we refactor via a separate issue?

File: schedule/scheduler.go (hunk 2 above)
Comment (iand) [reply schedule/scheduler.go:98]: Strong agree. Using os.Exit almost guarantees that any lock we manage to take will be left active in the database. So in some circumstances this exit behaviour will make our problem worse.

File: storage/migrations/17_chainvis_views_not_materialized.go (hunk 4 above)
Comment (frrist) [reply storage/migrations/17_chainvis_views_not_materialized.go:1]: As discussed in chat we will take these migration as is and flatten our schema https://github.com/filecoin-project/sentinel-visor/issues/225 in a follow on.

File: schedule/scheduler.go (hunk 1 above)
Comment (frrist) [reply schedule/scheduler.go:34]: As discussed in chat we will remove the `ExitOnFailure` and address the problem it is trying to fix in a follow on.

Comment (placer14) [reply schedule/scheduler.go]: As discussed, agreed to remove this cludgy fix entirely and find a better fix for considering `RestartOnCompletion` within the locking code.

File: schedule/scheduler.go (hunk 2 above)
Comment (placer14) [reply schedule/scheduler.go]: As discussed, agreed to remove here.

File: commands/run.go [hunk 5]
```diff
@@ -307,14 +307,12 @@ var Run = &cli.Command{
 			scheduler.Add(schedule.TaskConfig{
 				Name:                fmt.Sprintf("ActorStateChangeProcessor%03d", i),
 				Task:                actorstate.NewActorStateChangeProcessor(rctx.db, rctx.opener, cctx.Duration("statechange-lease"), cctx.Int("statechange-batch"), heightFrom, heightTo),
-				RestartOnFailure:    true,
-				RestartOnCompletion: true,
-				RestartDelay:        time.Minute,
```
Comment (frrist) [inline commands/run.go:312]: Note that this change will change the restart delay to 100 milliseconds which is probably reasonable(?)

**Review by frrist — APPROVED** [review]
:shipit: Thanks for wrangling this @placer14 :pray:

**Pull request merged**